hack/kind-cluster-build.sh

```diff
@@ -119,6 +125,137 @@ echo "start to create k8s cluster"
+    spec:
+      hostNetwork: true
+      nodeSelector:
+        kubernetes.io/hostname: ${registryNode}
+      tolerations:
+      - key: node-role.kubernetes.io/master
+        operator: "Equal"
+        effect: "NoSchedule"
+      containers:
+      - name: registry
+        image: registry:2
+        volumeMounts:
+        - name: data
+          mountPath: /data
+      volumes:
+      - name: data
+        hostPath:
+          path: /data
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: registry-proxy-config
+data:
+  nginx-conf: |
+    user  nginx;
+    worker_processes  1;
+    error_log  /var/log/nginx/error.log warn;
+    pid        /var/run/nginx.pid;
+    events {
+        worker_connections  1024;
+    }
+    http {
+        upstream docker-registry {
+            server ${registryNodeIP}:5000;
+        }
+        map \$upstream_http_docker_distribution_api_version \$docker_distribution_api_version {
+            '' 'registry/2.0';
+        }
+        server {
+            listen 5000;
+            server_name localhost;
+            client_max_body_size 0;
+            chunked_transfer_encoding on;
+            location /v2/ {
+                if (\$http_user_agent ~ "^(docker\/1\.(3|4|5(?!\.[0-9]-dev))|Go ).*$") {
+                    return 404;
+                }
+                add_header 'Docker-Distribution-Api-Version' \$docker_distribution_api_version always;
+                proxy_pass http://docker-registry;
+                proxy_set_header Host \$http_host;
+                proxy_set_header X-Real-IP \$remote_addr;
+                proxy_set_header X-Forwarded-For \$proxy_add_x_forwarded_for;
+                proxy_set_header X-Forwarded-Proto \$scheme;
+                proxy_read_timeout 900;
+            }
+        }
+        default_type  application/octet-stream;
+        log_format  main  '\$remote_addr - \$remote_user [\$time_local] "\$request" '
+                          '\$status \$body_bytes_sent "\$http_referer" '
+                          '"\$http_user_agent" "\$http_x_forwarded_for"';
+        access_log  /var/log/nginx/access.log  main;
+        sendfile        on;
+        keepalive_timeout  65;
+    }
+
+---
+apiVersion: extensions/v1beta1
+kind: DaemonSet
+metadata:
+  name: registry-proxy
+  labels:
+    app: registry-proxy
+spec:
+  selector:
+    matchLabels:
+      app: registry-proxy
+  template:
+    metadata:
+      labels:
+        app: registry-proxy
+    spec:
+      hostNetwork: true
+      affinity:
+        nodeAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: kubernetes.io/hostname
+                operator: NotIn
+                values:
+                  - ${registryNode}
+      tolerations:
+      - key: node-role.kubernetes.io/master
+        operator: "Equal"
+        effect: "NoSchedule"
+      containers:
+        - name: nginx
+          image: nginx:1.15-alpine
```

Comment:
register:2 has builtin http server, what's the purpose of nginx proxy?

Reply:
Because the kubelets use localhost:5000 to get images, so need to deploy `nginx-proxy` on every node.

Reply:
Only the localhost is trusted by Docker or containerd. Otherwise, we have to configure TLS for the registry.

Reply:
I see, this makes sense

Reply:
BTW, is a plain TCP proxy tunneling enough here? if it is, use socat (docker [image](https://hub.docker.com/r/alpine/socat/)) is much simpler, e.g. `socat tcp-listen:5000,fork,reuseaddr tcp-connect:<registry2-ip>:5000`.